### PR TITLE
Solana testnet automatic redeem support

### PR DESCRIPTION
**Pull request by kev1n-peters**

### CODE REVIEW

**Comment by kev1n-peters** [comment]
The contract interfaces may change after an internal audit, so putting this back into draft mode.

File: wormhole-connect/src/utils/routes/bridge.ts [hunk 1]
```diff
@@ -29,9 +29,18 @@ import {
   TransferDestInfoBaseParams,
   TransferInfoBaseParams,
 } from './types';
-import { hexlify } from 'ethers/lib/utils.js';
 import { isCosmWasmChain } from '../cosmos';
 import { fetchVaa } from '../vaa';
+import { hexlify } from 'ethers/lib/utils.js';
+
+export interface BridgePreviewParams {
```
Comment (M-Picco) [inline wormhole-connect/src/utils/routes/bridge.ts]: Should this interface be here? I don't see it used anywhere else

File: wormhole-connect/public/index.html [hunk 2]
```diff
@@ -15,7 +15,7 @@
   </head>
   <body style="margin: 0">
     <noscript>You need to enable JavaScript to run this app.</noscript>
-    <!-- <div id="wormhole-connect" config='{"networks": ["goerli", "mumbai"], "tokens": ["ETH", "WETH", "MATIC", "WMATIC"], "theme": "light"}'></div> -->
-    <div id="wormhole-connect"></div>
+    <div id="wormhole-connect" config='{"env": "testnet", "networks": ["fuji", "solana", "bsc", "goerli"], "tokens": ["AVAX", "WAVAX", "SOL", "WSOL", "BNB", "WBNB", "ETH", "WETH"], "theme": "light"}'></div>
+    <!-- <div id="wormhole-connect"></div> -->
```
Comment (M-Picco) [inline wormhole-connect/public/index.html]: Should this be commented out?

File: sdk/src/contexts/solana/relayer.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+      this.getRedeemerConfig(),
+    ]);
+    const swapRatePrecisionBN = new BN(swapRatePrecision);
+    const nativeSwapRate = this.calculateNativeSwapRate(
+      swapRatePrecisionBN,
+      solSwapRate,
+      swapRate,
+    );
+    const maxSwapAmountIn =
+      decimals > SOL_DECIMALS
```
Comment (M-Picco) [inline sdk/src/contexts/solana/relayer.ts:80]: Does it need to be decouple in two cases? If `decimals < SOL_DECIMALS` the exponent `decimals - SOL_DECIMALS` will be negative, which is the same as dividing by `TEN.pow(SOL_DECIMALS - decimals)`. Perhaps there's something from the BN lib I'm missing

File: sdk/src/contexts/solana/relayer.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        this.getRedeemerConfig(),
+      ]);
+    const swapRatePrecisionBN = new BN(swapRatePrecision);
+    const nativeSwapRate = this.calculateNativeSwapRate(
+      swapRatePrecisionBN,
+      solSwapRate,
+      swapRate,
+    );
+    const swapAmountOut =
+      decimals > SOL_DECIMALS
```
Comment (M-Picco) [inline sdk/src/contexts/solana/relayer.ts:108]: Same here

File: sdk/src/contexts/solana/relayer.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+    return BigInt(swapAmountOut.toString());
+  }
+
+  async fetchSwapEvent(signature: string): Promise<SwapEvent | null> {
+    const transaction = await this.connection.getParsedTransaction(signature);
+    if (transaction) {
+      const logMessages = transaction.meta?.logMessages || [];
+      for (const msg of logMessages) {
+        const matches =
```
Comment (M-Picco) [inline sdk/src/contexts/solana/relayer.ts:124]: Is there any other way to retrieve this data or is using a regex the only "elegant" way?

**Review by M-Picco — COMMENTED** [review]
Didn't notice anything out of place. Just some minor comments.

File: wormhole-connect/src/utils/routes/bridge.ts (hunk 1 above)
Comment (kev1n-peters) [reply wormhole-connect/src/utils/routes/bridge.ts]: This shouldn't be here, removed

File: wormhole-connect/public/index.html (hunk 2 above)
Comment (kev1n-peters) [reply wormhole-connect/public/index.html]: This shouldn't be changed, reverted

File: sdk/src/contexts/solana/relayer.ts (hunk 5 above)
Comment (kev1n-peters) [reply sdk/src/contexts/solana/relayer.ts:124]: Unfortunately this is how we have to do it on Solana. The contract calls `msg!` which is just a macro that logs data in the transaction.

File: sdk/src/contexts/solana/relayer.ts (hunk 3 above)
Comment (kev1n-peters) [reply sdk/src/contexts/solana/relayer.ts:80]: Good catch, we can simplify this

Comment (kev1n-peters) [reply sdk/src/contexts/solana/relayer.ts:80]: Actually, I changed this and it broke the gas slider. Here's a little example that shows that the two expressions aren't equivalent:

```
const TEN = new BN(10);
const SWAP_RATE_PRECISION = new BN(100_000_000);
const SOL_DECIMALS = 9;

function calculateSwapAmount(
  decimals: number,
  maxNativeSwapAmount: BN,
  nativeSwapRate: BN,
  firstCaseOnly: boolean = false
) {
  return decimals > SOL_DECIMALS || firstCaseOnly
    ? maxNativeSwapAmount
        .mul(nativeSwapRate)
        .mul(TEN.pow(new BN(decimals - SOL_DECIMALS)))
        .div(SWAP_RATE_PRECISION)
    : maxNativeSwapAmount
        .mul(nativeSwapRate)
        .div(TEN.pow(new BN(SOL_DECIMALS - decimals)).mul(SWAP_RATE_PRECISION));
}

  console.log(calculateSwapAmount(8, new BN(1000000), new BN(100000)));
  console.log(calculateSwapAmount(8, new BN(1000000), new BN(100000), true));

// prints
<BN: 64>
<BN: 2710>
```

Comment (kev1n-peters) [reply sdk/src/contexts/solana/relayer.ts:80]: So after some digging I've discovered that bn.js simply doesn't support `pow()` using negative numbers, e.g.
```
> new BN(10).pow(new BN(-2)).toString()
'100'
```

Comment (M-Picco) [reply sdk/src/contexts/solana/relayer.ts:80]: No choice but to leave it as is then

**Review by M-Picco — DISMISSED** [review]

**Review by M-Picco — APPROVED** [review]

**Pull request merged**